client/src/contexts/profile.js

```diff
@@ -1,43 +1,63 @@
-import React, {useContext, createContext } from 'react';
-import { withRouter } from 'react-router-dom';
+import React, {
+  useContext,
+  createContext,
+  useState,
+  useEffect,
+  useRef,
+} from "react";
+import { useLocation } from "react-router-dom";
 
 export const ProfileContext = createContext(null);
 
 export const useProfileContext = () => {
-    const profileContext = useContext(ProfileContext);
-    return profileContext;
-}
+  const profileContext = useContext(ProfileContext);
+  return profileContext;
+};
 
-class ProfileProvider extends React.Component {
-    constructor(props) {
-        super(props)
-        this.state = {
-            profile: null,
-            setProfile: profile => { this.setState({profile}) }
-        }
-    }
-    componentDidMount() {
-        fetch("/profile/me").then(profile => {
-            profile.json().then(result => {
-                this.state.setProfile(result);
-            })
-        })
-    }
-    componentDidUpdate(prevProps) {
-        const hasLoggedIn = (prevProps.location.pathname === "/login" || prevProps.location.pathname === "/signup") && (this.props.location.pathname !== "/login" || this.props.location.pathname !== "/signup")
-        if (hasLoggedIn) {
-            fetch("/profile/me").then((profile) => {
-              profile.json().then((result) => {
-                this.state.setProfile(result);
-              });
-            });
-        }
+function ProfileProvider(props) {
+  const [profile, setProfile] = useState(null);
+  const firstUpdate = useRef(true);
+  const prevLocation = useRef(null);
+  const context = {
+    profile,
+    setProfile: (profileUpdate) => {
+      const newProfile = { ...profileUpdate };
+      setProfile(newProfile);
+    },
+    pullProfile: () => {
+      fetch("/profile/me").then((profile) => {
+        profile.json().then((result) => {
+          setProfile(result);
+        });
+      });
     }
-    render() {
-        return (
-            <ProfileContext.Provider value={this.state}>{this.props.children}</ProfileContext.Provider>
-        )
+  };
+  const location = useLocation().pathname;
+  useEffect(() => {
+    if (firstUpdate.current) {
+      fetch("/profile/me").then((profile) => {
+        profile.json().then((result) => {
+          setProfile(result);
+        });
+      });
+      firstUpdate.current = false;
+    } else {
+      const hasLoggedIn = (prevLocation.current === "/login" || prevLocation.current === "/signup") && (location !== "/login" && location !== "/signup");
+      if (hasLoggedIn) {
+        fetch("/profile/me").then((profile) => {
```

Comment:
Chunk is repeated, is this accounting for eventually having non-protected homepages? If not, there's a case for moving this into its own function in case functionality in this section grows.